general/mem_manager.hpp

```diff
@@ -26,28 +26,64 @@ namespace mfem
 /// Memory types supported by MFEM.
 enum class MemoryType
 {
-   HOST,      ///< Host memory; using new[] and delete[]
-   HOST_32,   ///< Host memory aligned at 32 bytes (not supported yet)
-   HOST_64,   ///< Host memory aligned at 64 bytes (not supported yet)
-   CUDA,      ///< cudaMalloc, cudaFree
-   CUDA_UVM   ///< cudaMallocManaged, cudaFree (not supported yet)
+   HOST,           ///< Host memory; using new[] and delete[]
+   HOST_32,        ///< Host memory; aligned at 32 bytes
+   HOST_64,        ///< Host memory; aligned at 64 bytes
+   HOST_DEBUG,     ///< Host memory; protected while in device mode
+   HOST_UMPIRE,    ///< Host memory; using Umpire
+   HOST_MANAGED,   ///< Host memory; using *MallocManaged, *Free
+   DEVICE,         ///< Device memory; using *Malloc, *Free
```

Comment:
I think it is much cleaner (as interface) if instead of `DEVICE` we have separate `CUDA` and `HIP` types. Similarly for `DEVICE_MANAGED`.

Reply:
From a user/developer point of view, I don't see why we should be aware of each existing (and future) backend and their corresponding types.

The MFEM `Device` now offers two memory spaces: the `Host` and `Device` ones, and the user/developer should only care about these.

Shouldn't this distinction be sufficient, as long as we don't plan to support heterogeneous devices simultaneously?

Reply:
My thinking was that at some point it may be useful to support CUDA and HIP in the same build. For example, on Lassen we may be able to mix them (with the hip backend targetting the nvidia hardware).

@tzanio, @dmed256, what do you think? Should we have separate `MemoryType::CUDA` and `MemoryType::HIP` or just `MemoryType::DEVICE`?

Reply:
My vote is for `MemoryType::DEVICE`. I don't see an important use case where we will need them both in the same build.

Reply:
OK, let's document this -- add to the doxygen comment that `DEVICE` means either `CUDA` or `HIP`; similarly, add a comment for the `MANAGED` type.

Also, it will be good to mention this interface change in CHANGELOG.

Reply:
`done`